src/CSV.jl

```diff
@@ -364,141 +380,138 @@ function file(source,
+    # if a column type if promoted to string, the values are stored in the corresponding `tape` instead of `poslen`
     pool = pool === true ? 1.0 : pool isa Float64 ? pool : 0.0
+    intsentinel = Ref{Int64}(1)
     if threaded === true
         # multithread
-        rows, tapes, refs, typecodes = multithreadparse(typecodes, buf, datapos, len, options, rowsguess, pool, ncols, ignoreemptylines, typemap, limit, cmt, debug)
+        rows, tapes, refs, typecodes = multithreadparse(typecodes, buf, datapos, len, options, rowsguess, pool, ncols, ignoreemptylines, typemap, limit, cmt, intsentinel, debug)
+        finalrows = sum(rows)
     else
-        tapelen = rowsguess * 2
-        tapes = Vector{UInt64}[Mmap.mmap(Vector{UInt64}, tapelen) for i = 1:ncols]
+        tapes, poslens = allocate(rowsguess, ncols, typecodes)
         refs = Vector{Dict{String, UInt64}}(undef, ncols)
         lastrefs = zeros(UInt64, ncols)
         t = Base.time()
-        rows, tapes = parsetape(Val(transpose), ignoreemptylines, ncols, gettypecodes(typemap), tapes, tapelen, buf, datapos, len, limit, cmt, positions, pool, refs, lastrefs, rowsguess, typecodes, debug, options, false)
+        rows, tapes = parsetape(Val(transpose), ignoreemptylines, ncols, gettypecodes(typemap), tapes, poslens, buf, datapos, len, limit, cmt, positions, pool, refs, lastrefs, rowsguess, typecodes, intsentinel, debug, options, false)
+        finalrows = rows
         debug && println("time for initial parsing to tape: $(Base.time() - t)")
     end
     for i = 1:ncols
         typecodes[i] &= ~USER
     end
     finaltypes = Type[TYPECODES[T] for T in typecodes]
     debug && println("types after parsing: $finaltypes, pool = $pool")
-    finalrefs = Vector{Vector{String}}(undef, ncols)
+    finalrefs = Vector{Union{Vector{String}, Nothing}}(undef, ncols)
     if pool > 0.0
         for i = 1:ncols
             if isassigned(refs, i)
                 finalrefs[i] = map(x->x[1], sort!(collect(refs[i]), by=x->x[2]))
-            elseif typecodes[i] == POOL || typecodes[i] == (POOL | MISSING)
+            elseif typebits(typecodes[i]) == POOL
                 # case where user manually specified types, but no rows were parsed
                 # so the refs never got initialized; initialize them here to empty
                 finalrefs[i] = Vector{String}[]
+            else
+                finalrefs[i] = nothing
             end
         end
+    else
+        fill!(finalrefs, nothing)
     end
-    return File(getname(source), names, finaltypes, rows, ncols, eq, categorical, finalrefs, buf, tapes)
+    if threaded === true
+        columns = AbstractVector[ApplyArray(vcat, (Column{_eltype(finaltypes[i]), finaltypes[i]}(tapes[j][i], rows[j], eq, categorical, finalrefs[i], buf, finaltypes[i] >: Int64 ? uint64(INT_SENTINELS[intsentinel.x]) : sentinelvalue(Base.nonmissingtype(finaltypes[i]))) for j = 1:Threads.nthreads())...) for i = 1:ncols]
```

Comment:
For the multithreaded case, each thread parses an equal chunk of the file. Previously, we were then allocating a final, huge tape and copying each thread's chunk into the final tape.

Here, we're using the quite-well-optimized `LazyArray.Vcat` type to lazily `vcat` the threads' tapes into a final "column". So essentially in the non-threaded case, you get back a `Vector{Column}`, and in the multithreaded case, you get back `Vector{LazyArrays.Vcat{Column}`. I've tried to do a pretty thorough benchmarking/stress-testing on `LazyArray.Vcat` and I've been really impressed with the optimizations they've implemented for broadcasting, reductions, etc. 

I also feel like we've already conditioned users to know that in the `copycols=false` case, you're getting some kind of optimized "view" `AbstractVector`, and to use `copycols=true` if they need a final, full `Vector`.